### ISSUE TITLE
Does C.21 apply to abstract classes?

### ISSUE THREAD

**Issue by sweemer** [isocpp/CppCoreGuidelines#1897]
[C.21: If you define or `=delete` any copy, move, or destructor function, define or `=delete` them all](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all) currently covers the following cases:

- class with non-virtual destructor
- class with virtual destructor but no pure virtual functions

However it doesn't cover the case of abstract classes - that is, classes containing both a virtual destructor and one or more pure virtual functions. In this case, is it still desirable to enforce the Rule of Five?

As abstract classes cannot be move-constructed, move-assigned, copy-constructed, or copy-assigned, it seems a bit pedantic to require the class to define those four special member functions. Instead, the Rule of Five should apply to concrete classes only.

I propose that C.21 be renamed to _If you define or `=delete` any copy, move, or destructor function **in a concrete class**, define or `=delete` them all_.

Also, as a side note, the `AbstractBase` example in C.21 is not actually an abstract class because it does not define any pure virtual functions. I propose that it be renamed to just `Base` or something similar to avoid potential confusion.

**Comment by MikeGitb** [comment]
While and abstract class can't be constructed on its own it may be constructed as part of a derived object. 
Concrete example: 9 times out of 10, the best way to implement a clone function is to use a private/protected copy constructor, which in turn is much easier to implement when the base to have a protected copy constructor too.

So yes, I absolutely believe the rule of 0/5 should apply to abstract classes as well.

**Comment by shaneasd** [comment]
As for assignment, you can also assign to a reference to an abstract class which could give you some pretty weird behavior as in this example where you end up with the data of C but the virtual function table of B.
https://godbolt.org/z/x7ob6qxG7

This rule would seem to be in agreement with https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c67-a-polymorphic-class-should-suppress-public-copymove so it seems to me that if you think an exception is needed for abstract classes here then you would also disagree (rightly or wrongly) with C.67.

I suppose you could argue that this should be mentioned as a special case for C.21 in that the rules is more specific than implement, default or delete them in that C.67 is always invoked in this case meaning you have to also make them protected.

**Comment by shaneasd** [comment]
Thanks. I always forget that evil extension in msvc. For clarity https://godbolt.org/z/v9MGWP9hb is essentially the same code without that problem.

**Comment by sweemer** [comment]
> So yes, I absolutely believe the rule of 0/5 should apply to abstract classes as well.

Strictly speaking, the Rule of Zero should not apply to abstract classes, because of _C.35: A base class destructor should be either public and virtual, or protected and non-virtual_. In both cases, the base class destructor must be defined, and therefore the Rule of Five only must be followed.

> it seems to me that if you think an exception is needed for abstract classes here then you would also disagree (rightly or wrongly) with C.67.

As far as I understand, C.67 deals with polymorphic classes, not abstract classes, the difference being of course that abstract classes contain at least one _pure_ virtual function. In the case of polymorphic classes, I agree that the risk of slicing warrants full adherence to the Rule of Five, but in the case of abstract classes, there is no risk of slicing and so I don't think that part of C.67 applies.

As another side note, the "good" example in C.67 does not follow C.21 because the move constructor and move assignment operator are omitted. Those should probably be added to the example.

I actually thought of a good reason why the Rule of Five should still apply to abstract classes - when the abstract class contains a member variable that requires resource management in the destructor, then the copy and move special functions should not be exempted from C.21 simply because the class is abstract.

So in conclusion I believe that C.21 does apply to abstract classes after all, and that the following point in C.67 provides the correct approach for how to apply C21 to abstract classes: _If the class has no data, =delete the copy/move functions. Otherwise, make them protected_.

**Comment by MikeGitb** [comment]
> Strictly speaking, the Rule of Zero should not apply to abstract classes, because of C.35: A base class destructor should be either public and virtual, or protected and non-virtual. In both cases, the base class destructor must be defined, and therefore the Rule of Five only must be followed.

An abstract class isn't necessarily the root of the class hierarchy. As long as the root declares its destructor virtual, all destructors in derived classes are virtual, even if they follow the rule of zero.

> I agree that the risk of slicing warrants full adherence to the Rule of Five, but in the case of abstract classes, there is no risk of slicing and so I don't think that part of C.67 applies.

The risk might be reduced, but not zero:

    void foo(AbstractBase& a, const AbstractBase& b) 
    {
       a = b;
    }
    
> If the class has no data, =delete the copy/move functions. Otherwise, make them protected.

I actually don't understand, why the presence of member variables should have an effect on whether the SMF should be protected or deleted, as thats an interface decision, but I guess that would be a topic for a different issue.

**Comment by sweemer** [comment]
> I actually don't understand, why the presence of member variables should have an effect on whether the SMF should be protected or deleted, as thats an interface decision, but I guess that would be a topic for a different issue.

From an interface perspective, the end result of following this part of C.67 is the same - public copy/move is disallowed. The difference between the two cases is that `=delete` is the clearest way to disallow copy/move for a polymorphic base class containing no data, but one containing some data may need to define the appropriate copy/move SMFs in order to facilitate delegation from the corresponding derived class SMFs.

**Comment by MikeGitb** [comment]
> From an interface perspective, the end result of following this part of C.67 is the same - public copy/move is disallowed. 

No it is not the same, because in one case no one can use a copy constructor and in the other derived classes can use the copy constructor. As the derived class (usually) shouldn't have to care if the base class has internal state or not, I don't see a reason to reflect this implementation detail in the base class interface. Only if I don't want the base class (and derived ones) to be copyable at all does it imho make sense to delete the copy constructor. But that is (imho) an othogonal issue to the presence or absence of state in the base class.
Point in case: Pretty much every (rgeular) statless class I've ever written was (trivially) copyable.
